### ISSUE TITLE
User can still approve their own changes

### ISSUE THREAD

**Issue by JakeKerber** [MicrosoftDocs/vsts-docs#1198]
I believe I ran into a situation in which a developer can approve their own changes even if "Allow users to approve their own changes" is unchecked - assuming only 1 approver is required.
1) Alice creates a pull request for BranchA (containing her new code) to be merged into master.
2) Bob pulls down BranchA and makes a new change to it and pushes it up. In our situation this resets any reviewer votes on the pull request since we have that option enabled in our branch policy on master.
3) Bob can now approve the pull request. Only Alice is prevented from approving the pull request she created - even though it also contains changes from another developer.

Am I understanding this correctly? Is there a way to prevent this (besides requiring a 2nd reviewer)?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7cd986e0-608c-1d8c-6faf-ecdb12e10bc6
* Version Independent ID: d60c996e-187e-dd6b-7f3a-dd056a1603f4
* Content: [Protect your Git branches with policies](https://docs.microsoft.com/en-us/vsts/git/branch-policies?view=vsts)
* Content Source: [docs/git/branch-policies.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/git/branch-policies.md)
* Product: **devops**
* GitHub Login: @steved0x
* Microsoft Alias: **sdanie**

**Comment by steved0x** [comment]
@JakeKerber Let me check on that for you. I just did that scenario and it is like you described:

- Alice can't approve the PR (and have it count towards the minimum number of required reviewers policy) because she is the PR creator and "Allow users to approve their own changes" is not checked, even if the PR contains pushes from other developers (Bob). 
- However, Bob can approve the PR and complete it without the requirement for another reviewer to approve, therefore "approving his own changes"

You would like to prevent the 2nd issue - Bob approving and completing the PR with his changes without having a 2nd reviewer required?

Having 2 minimum required reviewers would resolve that, because you have the votes set to reset on new pushes, which would mean that Bob and Alice would need to review and approve after each new change was pushed, which would give the desired effect of having a review by someone other than themselves. (And actually a 3rd person since Alice's vote doesn't count to meet that requirement), but you are asking, how to prevent that without upping the count to 2?

Possibly specifying a group as a required reviewer, but if Bob was in the group he could still approve his own changes if he wasn't the PR opener.

Let me see what I can find out for you.

**Comment by mmitrik** [comment]
Thanks for looping me in @steved0x 

Yes, this scenario is currently a gap, and we've been discussing ideas for how we might remediate it.  In the short term, I think the only way to prevent this is to require a 2nd reviewer.  

While we're on the subject, two options we've considered here are:

1. Change the behavior of the "Allow users to approve their own changes" option to exclude all users that have contributed changes to the PR.  This would include the PR author, as well as the anyone that pushed a commit (commit authors and committers are unreliable).  This is the most restrictive option, and kills scenarios like pair-programming where two people collaborate on a single PR.  

2. Change the "Allow users to approve their own changes" option to "Allow the last person to push changes to approve the pull request".  If both Alice and Bob approve a PR where Alice was the last person to change the source branch, then only Bob's vote would count.  If Bob then pushed changes, only Alice's vote would count (assuming the reset votes option wasn't enabled).  This is less restrictive, but certainly more confusing to the casual observer, as the votes that "count" would flip-flop.  

Neither of those seem ideal, and adding an additional reviewer seems to unblock most teams.  This is, however, something we should address.

**Comment by steved0x** [comment]
@JakeKerber I've also added a clarification in this article about who the `user` is in `Allow users to approve their own changes` to clarify that `user` in this policy is the pull request creator - that should be live in a day or two.

I'm going to close this doc feedback issue, and to continue the discussion further with @mmitrik and the rest of the team, please file a feature request at https://developercommunity.visualstudio.com/spaces/21/index.html and share whether you like approach 1 or 2 above, or if you had another idea that would work better for your team.

Thanks again for the feedback and great question!
